### PR TITLE
fix: run ingress filter in StateMachine tests

### PR DESCRIPTION
**Pull request by mraszyk**
This MR runs ingress filter in StateMachine tests to avoid critical errors of the form `execution_environment_unfiltered_ingress`.

The [ingress filter](https://github.com/dfinity/ic/blob/2a2b74684b55c40d514c2e80165432e906dddbf5/rs/execution_environment/src/execution_environment.rs#L2307) performs a bunch of checks for every ingress message before this ingress message is accepted for execution. In particular, the checks include:
- `canister_status` on the management canister can only be called by controllers (this does not include the canister itself requesting its own status via an _ingress_ message as a canister would request its status via an _inter-canister_ call in production); hence, the controllers of canisters with no controllers can only be retrieved via a `canister_info` call proxied through an intermediate canister;
- ingress messages cannot request the `canister_info` and `fetch_canister_logs` endpoint;
- ingress messages to empty/stopped canisters are rejected.

### CODE REVIEW

File: rs/nns/integration_tests/src/cycles_minting_canister.rs [hunk 1]
```diff
@@ -514,8 +542,8 @@ fn test_cmc_cycles_create_with_settings() {
         10_000_000_000_000,
     )
     .unwrap();
-    let status = canister_status(&state_machine, *TEST_USER1_PRINCIPAL, canister);
```
Comment (max-dfinity) [inline rs/nns/integration_tests/src/cycles_minting_canister.rs:517]: Why is this change necessary?

File: rs/nns/test_utils/src/state_test_helpers.rs [hunk 2]
```diff
@@ -361,7 +361,7 @@ pub fn get_canister_status(
 
 pub fn get_root_canister_status(machine: &StateMachine) -> Result<CanisterStatusResultV2, String> {
     machine
-        .canister_status_as(PrincipalId::from(ROOT_CANISTER_ID), ROOT_CANISTER_ID)
+        .canister_status_as(PrincipalId::from(LIFELINE_CANISTER_ID), ROOT_CANISTER_ID)
```
Comment (max-dfinity) [inline rs/nns/test_utils/src/state_test_helpers.rs:364]: Can root not request its own canister status?

File: rs/nns/integration_tests/src/cycles_minting_canister.rs (hunk 1 above)
Comment (max-dfinity) [reply rs/nns/integration_tests/src/cycles_minting_canister.rs:517]: To add context, I guess I'm asking - what's changing about the state machine API?

**Review by max-dfinity — APPROVED** [review]

File: rs/nns/test_utils/src/state_test_helpers.rs (hunk 2 above)
Comment (mraszyk) [reply rs/nns/test_utils/src/state_test_helpers.rs:364]: It can't do so via an _ingress_ message since in production a canister would request its own status via an _inter-canister_ call. Now this PR adds the ingress filtering as performed in production which prevents calls to `canister_status` from non-controllers.

File: rs/nns/integration_tests/src/cycles_minting_canister.rs (hunk 1 above)
Comment (mraszyk) [reply rs/nns/integration_tests/src/cycles_minting_canister.rs:517]: I've updated the PR description to provide more background on why the changes of the tests are necessary. In this particular case, we have a canister with no controllers and thus the `canister_status` call fails. Before this PR, the error message would come from the execution of the message and it would contain an explicit list of canister controllers whose content was asserted in the test. With the change from this PR, the error message comes from the ingress filter (as it'd do in production) and thus it is less verbose and the set of controllers cannot be derived from the new error message => we use the `canister_info` endpoint that can be called by an arbitrary (proxy) canister.

File: rs/execution_environment/tests/dts.rs [hunk 3]
```diff
@@ -351,20 +352,65 @@ fn dts_install_code_with_concurrent_ingress_sufficient_cycles() {
...
+            wat::parse_str(DTS_INSTALL_WAT).unwrap(),
+            vec![],
+            None,
+            None,
+        )
+        .encode(),
+    )
+    .unwrap();
+
+    let install_code_ingress_id = env.send_ingress(
```
Comment (berestovskyy) [inline rs/execution_environment/tests/dts.rs:400]: Why do we need to install the canister first?

Comment (mraszyk) [reply rs/execution_environment/tests/dts.rs:400]: Ingress messages to empty canisters are rejected by the ingress filter before they are enqueued. Hence, we need to make the canister non-empty so that the ingress message passes the ingress filter and gets enqueued for processing after the DTS completes.

**Review by alin-at-dfinity — APPROVED** [review]

**Review by berestovskyy — APPROVED** [review]

File: rs/nns/test_utils/src/state_test_helpers.rs (hunk 2 above)
Comment (max-dfinity) [reply rs/nns/test_utils/src/state_test_helpers.rs:364]: I'm wondering if state_machine could eventually have a feature that adds the ability to send inter-canister calls directly?  That was a great feature that allowed us to avoid using the universal canister when doing the testing.

Comment (mraszyk) [reply rs/nns/test_utils/src/state_test_helpers.rs:364]: > I'm wondering if state_machine could eventually have a feature that adds the ability to send inter-canister calls directly? That was a great feature that allowed us to avoid using the universal canister when doing the testing.

The only easy option I see is to have the StateMachine test API deploy a universal canister behind the scenes and proxy the call through it. But then the caller of the proxied call would be that universal canister which might be an issue in the case of mgmt canister calls.

A more involved approach might be to enqueue an actual inter-canister call as if it was submitted during an actual canister message execution, specifying -1 as the callbacks and marking the parent call context as deleted. But this sounds pretty ad-hoc and I'm not sure if this wouldn't violate other assumptions in the system.

Comment (max-dfinity) [reply rs/nns/test_utils/src/state_test_helpers.rs:364]: Hm - I'm not sure if this is actually a good change to make to our testing tools, if we are going to lose the ability to just make simple calls as other canisters.  It may make StateMachine more faithful to the replica, but is that faithfulness important at the level of testing canister APIs?

Comment (mraszyk) [reply rs/nns/test_utils/src/state_test_helpers.rs:364]: > if we are going to lose the ability to just make simple calls as other canisters

you're not losing that ability completely, only in cases where the ingress filter would reject such a message

Comment (mraszyk) [reply rs/nns/test_utils/src/state_test_helpers.rs:364]: And, e.g., ingress messages to the `canister_info` endpoint are already rejected even without this PR.

Comment (max-dfinity) [reply rs/nns/test_utils/src/state_test_helpers.rs:364]: > you're not losing that ability completely, only in cases where the ingress filter would reject such a message

Okay, then that's probably fine.  As long as we can still send a message from Canister principal, this is okay.
